functional-samples/sample.dnr-rule-manager/manager.js

```diff
@@ -0,0 +1,199 @@
+const ruleItemTemplate = document.getElementById('ruleItemTemplate');
+const rulesList = document.getElementById('rulesList');
+const addRuleButton = document.getElementById('addRuleButton');
+const viewRuleButton = document.getElementById('viewRuleButton');
+
+async function getNextRuleId() {
+  const rules = await chrome.declarativeNetRequest.getDynamicRules();
+  return Math.max(...rules.map((rule) => rule.id)) + 1;
+}
+
+async function refresh() {
+  const rules = await chrome.declarativeNetRequest.getDynamicRules();
+  renderRules(rules);
+}
+
+refresh();
+
+addRuleButton.addEventListener('click', () => {
+  appendRuleItem(rulesList, {
+    id: 'NEW',
+    conditionType: 'urlFilter',
+    conditionValue: '',
+    caseSensitive: true
+  });
+});
+
+viewRuleButton.addEventListener('click', async () => {
+  const rules = await chrome.declarativeNetRequest.getDynamicRules();
+  const rulesString = JSON.stringify(rules, null, 2);
+  const newWindow = window.open();
+  newWindow.document.write(
+    `<pre style="font-size: 1rem;">${rulesString}</pre>`
+  );
+});
+
+function appendRuleItem(
+  container,
+  { id, conditionType, conditionValue, caseSensitive }
+) {
+  // Create a new rule item from the template
+  const ruleItem = ruleItemTemplate.content.cloneNode(true).children[0];
+  container.appendChild(ruleItem);
+
+  // Set the rule ID
+  ruleItem.querySelector('.rule-id').value = id;
+
+  // Set the conditionType select
+  ruleItem.querySelector('.condition-type').value = conditionType;
+
+  // set the caseSensitive input
+  ruleItem.querySelector('.case-sensitive').checked = caseSensitive;
+
+  // Set the conditionValue input
+  ruleItem.querySelector('.condition-value').value = conditionValue;
+
+  // Set button click handlers
+  function getCurrentRuleValues() {
+    const _conditionValue = ruleItem.querySelector('.condition-value').value;
+    const _conditionType = ruleItem.querySelector('.condition-type').value;
+    const _caseSensitive = ruleItem.querySelector('.case-sensitive').checked;
+    return {
+      conditionValue: _conditionValue.trim(),
+      conditionType: _conditionType,
+      caseSensitive: _caseSensitive
+    };
+  }
+  ruleItem.querySelector('.remove-rule').addEventListener('click', async () => {
+    // For a new rule, just remove the item from the DOM
+    if (id !== 'NEW') {
+      await removeRule(id);
+    }
+    ruleItem.remove();
+  });
+  ruleItem.querySelector('.save-rule').addEventListener('click', async () => {
+    const {
+      conditionValue: _conditionValue,
+      conditionType: _conditionType,
+      caseSensitive: _caseSensitive
+    } = getCurrentRuleValues();
```

Comment:
Do we need to rename these? It seems like we shouldn't have to, which should hopefully clean up the code quite a bit.